### ISSUE TITLE
datetest appears to accept nonsense when using --isvalid

### ISSUE THREAD

**Issue by Earnestly** [hroptatyr/dateutils#146]
```sh
datetest --isvalid -i %FT%T -i %F -- 2022-02-27 # Okay
datetest --isvalid -i %FT%T -i %F -- 2022-02-27T # Should fail
datetest --isvalid -i %FT%T -i %F -- 2022-02-27T9999:8888:7777 # Should definately fail
```

It seems that if `2022-02-27` is matched, anything after it is also accepted.

Is there a proper way to have `datetest` check the validity of an input based on multiple formats (i.e. a subset of ISO8601 or RFC3339)?

**Comment by hroptatyr** [comment]
Hi,
thanks for the report, good catch.  Fix is in 06198a860.

**Comment by AleksejK1** [comment]
0.4.11 also validates on Feb dates after 28:
```sh
$ datetest --isvalid 2026-02-28; echo $?
0
$ datetest --isvalid 2026-02-29; echo $?
0
$ datetest --isvalid 2026-02-30; echo $?
0
$ datetest --isvalid 2026-02-31; echo $?
0
$ datetest --isvalid 2026-02-32; echo $?
1
```

**Comment by hroptatyr** [comment]
Hey Aleksej,
`--isvalid` checks for conformance to the input format string, as in: is this string valid input to the parser when using that input format.  So your examples behave as expected.

For why dateutils accept dates like 2026-02-31 or 2026-06-31: It's a convenient way to refer to the last day in a month, emit strings like %Y-%m-31 and they will be read and represented correctly internally.

This convenience is nearly everywhere, you can pass datetimes like `2026-03-29T02:05:00 --from-zone Europe/Berlin`, or `2026-06-31T07:04:60`, or something known as military midnight like `2026-03-29T24:00:00` (the latter two even survive in the output).

If you want to check whether some input would also classify as valid output you can use `dateconv` and compare the input to the output:

    $ dateconv 2026-02-31
    2026-02-28
    $ dateconv 2026-03-29T02:05:00 --from-zone Europe/Berlin --zone Europe/Berlin
    2026-03-29T01:05:00

but like in the examples above, sometimes the convenience makes its way into the output too:

    $ dateconv 2026-02-00
    2026-02-00

**Comment by AleksejK1** [comment]
Hi Sebastian,
Without going into dateutils' internals you explained, just from my point of
view as a user:

If "datetest --isvalid" does not *test* if the *date* *is valid* then it claims
(through its wording) to be more than it is.
"datetest --isproperlyformatteddatestring" would set the right expectations at
least.

Using dateconv and comparing outputs looks like a hack and adds code
complexity. Would be nice if datetest --isvalid (or any other single
command) actually tested for valid dates.

**Comment by hroptatyr** [comment]
I suppose it's easy for recent ISO 8601 YMD and YD dates in the UTC time zone.
However, ISO 8601 YWD dates, or YMCW dates are explicitly not unique in their representation, and I'd hate to introduce something that only works for a few select date notations.

May I asked what your use case is?

**Comment by AleksejK1** [comment]
My use case is to ask a user of the shell script to input a date. If date is
invalid - report error, ask again.

**Comment by hroptatyr** [comment]
And what's the allowed input format string and the allowed input range?  As it stands `datetest --isvalid` exclusively tests for validity of the input with respect to the input format and the supported range.

Like I said, if you have a very narrow idea of what dates you want to accept, like only ISO 8601 UTC dates, the easiest way to check is if the output representation matches the input representation.  That's quite easy to implement in a shell script.

**Comment by hroptatyr** [comment]
on the other hand, 1900-04-4 would be invalid according to that.

In general, though, if the date is processed with dateutils 2000-06-00 is just as valid as 2000-06-31, and there should be no complications.  If the date is to be passed to other tools (like a SQL db), then you need to check using their validity tools, 2000-06-00 is valid whereas 2000-06-31 is not.

**Comment by AleksejK1** [comment]
Allowed input range - The Past, including current time :) "Since 1970" is fine. Allowed input format - ISO 8601, but local time for users' convenience (I'm converting it to UTC later).

"easiest way to check is if the output representation matches the input representation" - I'm not sure what that means, could you elaborate?

**Comment by AleksejK1** [comment]
"processed with dateutils 2000-06-00 is just as valid as 2000-06-31" - in the real world usage they are both invalid though :) That's the complication. Coreutils' date recognizes this:

$ date --date='2000-06-00 16:00:00'
date: invalid date ‘2000-06-00 16:00:00’
$ date --date='2000-06-31 16:00:00'
date: invalid date ‘2000-06-31 16:00:00’

"If the date is to be passed to other tools (like a SQL db), then you need to check using their validity tools" - that's exactly my use case actually, but I wanted to safeguard *before* passing data to db.

**Comment by hroptatyr** [comment]
when I do things like that, I'll pass everything through dateconv -E.  Stuff that cannot be read is an empty row, everything else is in normal form.  Depending on the data vendor I might need a sed-massage, some use the rather esoteric 2000-05-99 to mean a year-month date with no explicit day but since it's rare (and not ISO compliant (whereas 2000-05-00 is)) I'm okay with that.

**Comment by hroptatyr** [comment]
Btw, if you're happy with coreutils' date, it has date -f to process a whole file.  On the other hand they only allow for output formatting, not input formatting, so maybe you go through

    $ cat stuff | date -Idate -f - | dateconv -Ei '%F'

**Comment by hroptatyr** [comment]
On the input = output test, like I said a very strict validator, I was thinking of:

    $ test "%x" = "`dateconv -i %F $x 2>/dev/null`"

Or in batch mode:

    $ diff -q $FILE <(dateconv -Ei %F < $FILE)

**Comment by AleksejK1** [comment]
Hacks, hacks and more hacks :)

"datetest --isvalid" would be beautifully concise and English.

Thanks for the tips though, I understand you don't want to change your
code, I'll make do with date -d 'stuff' > /dev/null

**Comment by hroptatyr** [comment]
Well I can't change the behaviour of --isvalid, as it is specifically to determine whether the input would be able to be parsed by the dateutils tools.

Furthermore, it's a major undertaking for a very specific use case when there's a simple(r) solution.  The thing is that dates like 2000-06-31, 2000-02-30, 2000-02-00, etc. have a distinct internal representation whose output happens to coincide with other representations.

In the upcoming 0.5.0 release, dateutils will support other calendric systems where months don't have fixed lengths (like the Hijri calendars).  The current design decision of keeping the input separate from the output (and not normalise values internally) has certainly helped implement this with no major overhauls.

**Comment by hroptatyr** [comment]
Hi Aleksej,
as it turns out, the hard bits I was too afraid to flesh out had been done already.  So I present 0aee6962, a datetest(1) that should suit you (apart from still allowing ISO compliant Y-M-00 dates).

**Comment by AleksejK1** [comment]
Hi Sebastian,
Tested it (the binary is called dtest not datetest for some reason) -
works as expected, thanks. Waiting for next release.

Duck.ai and Wikipedia (both admittedly shoddy sources) tell me that
Y-M-0 dates are not allowed by ISO 8601. Is there a source for this compliance?

**Comment by hroptatyr** [comment]
Yea, it's make install that renames the binaries.  Use `make install DESTDIR=$(pwd)` to put the installation into the build tree.

It's ISO 8601-1:2019 3.1.3.7 which handles omissions, and ISO 8601-2:2019 which names extensions.

It's a way to keep parsers happy that would otherwise run into context problems, like --05 or --05-- could be mistaken for a comment in some languages (SQL, lua, to name two).  So it allows replacements of omitted "fields" by out-of-range symbols, 0000-05 or XXXX-05 or ????-05 to name May in any year, and 0000-05-00 or 0000-05-XX to mean any day in May in any year.

If need be, dateutils only accept the 00 variants, and only omissions to the right, i.e. 2000-00-00 is okay, 2000-00-04 is not.  And omitting a year to the right, means you don't have a date, so 0000-00-00 is unsupported.  The SQL engines I work with can dates with omitted components, but if need be a simple grep -v for `-00` should also do the job.

Also, non-ISO dates will make it through: 2000-1-1 for one, 2000/1/1 for another.  THe latter could be caught by using explicity input formatting (`datetest --isvalid -i "%Y-%m-%d"` 2000/1/1), the former cannot be caught by input formatting trickery.

**Comment by AleksejK1** [comment]
I'll take your word for it.

FWIW, date -d accepts '2025-02-28 10:00:00f' as valid date and SQLite
accepts Feb days up to 31. So I'll be waiting for dateutils 0.4.12.

TL;DR: date parsing/validation is a mess :)

**Comment by hroptatyr** [comment]
Ah, tell me about it.  And to be fair, I see the point.  In finance (where I live) there's the act/360 calendar where every month has 30 days, and the year has 360 days.  So it's not uncommon to see dates like 2026-02-30, and given the stake of finance in about every industry, I imagine concepts like these make their way into standards and software quite easily.

In the case of sqlite I'd say it's ISO 9899 (the C standard) that mandates that the day-of-month field can take values [1,31] irrespective of the month field.
